### ISSUE TITLE
Accept hyphens in config key given to CLI

### ISSUE THREAD

**Issue by cmkobel** [snakemake/snakemake#2986]
TL;DR Can we allow hyphens in config keys given on the command line?

I'm designing a distributed pipeline consisting of many different bioinformatic tools. In order to make it easy to use, I wish to be able to address any possible command line argument to individual rules when calling snakemake on the command line. And that is by using the built in config command line parser in Snakemake.

I have already implemented a system that looks for a specific syntax in the config key and sends that key-value pair to a specific rule. The system works by looking for config keys following a specific syntax that allows addressing rule, key and value for the command line argument: "set_\<rule>\<key>=\<value>". 
Here "set_" is a static prefix; \<rule> is the Snakemake rule; \<key> is the command line option key; and \<value> is the argument value. By setting these in the config.yaml-file, one can easily configure unlimited settings for individual rules in the pipeline. For example, you might want to set the "--kingdom archaea" command line argument to the prokka rule when analysing archaeal genomes. In this case you would write "set_prokka--kingdom: archaea" in the config.yaml and the parser will then call prokka with `prokka --kingdom archaea [...]` automatically. 

The problem arises when I then try to set these on the command line. E.g. `snakemake --snakefile [...] --config set_prokka--kingdom=archaea`. I then get the error: "Invalid config definition: Config entry must start with a valid identifier." This is because [Snakemake's parse_config() command line config parser](https://github.com/snakemake/snakemake/blob/e8735c1477a2a82110757ba86bbd1ccbcaf327ba/snakemake/cli.py#L248) uses the regex string "[a-zA-Z_]\w*$" to validate the keys given to the config, which doesn't allow hyphens in the config keys _at all_. 

My proposed workaround (https://github.com/cmkobel/snakemake/commit/94794189eb60c1e3d1bdce61abf10372b79270d6) is to change this regex string to "[a-zA-Z_][\w-]*\w$", thereby allowing hyphens in the middle of config keys (when given on the command line). I already implemented this in my pipeline as a proof-of-concept. It works as intended.

Unfortunately it is not sustainable that I have to [tell the users of my pipeline](https://comparem2.readthedocs.io/en/latest/20%20usage/#passthrough-arguments) to change the internal code of Snakemake though this can easily be done with the following sed command:

```bash
sed -i 's|valid = re.compile(r"\[a-zA-Z_\]\\w\*\$")|valid=re.compile(r"\[a-zA-Z_\]\[\\w-\]\*\\w\$")|g' $CONDA_PREFIX/lib/python3.11/site-packages/snakemake/__init__.py
```

So with this feature request, I'd like the Snakemake developers to consider if we can change the regex string of [parse_config()](https://github.com/snakemake/snakemake/blob/e8735c1477a2a82110757ba86bbd1ccbcaf327ba/snakemake/cli.py#L248) to allow hyphens in the config keys.  For example with r"[a-zA-Z_][\w-]*\w$" instead.

By allowing Snakemake to accept hyphen-containing keys for the config on the command line, the usability of the built in command line parser is greatly enhanced. As hyphens are already valid in keys given in the config-file, I don't see any obvious problems to allow these on the command line.

Just want to humbly brainstorm this with you before making a pull request. 

And thanks for developing a great framework. Best, Carl

**Comment by eternal-flame-AD** [comment]
My opinion is I don't think there is a necessity to change this:

Firstly the reason that regex was written like that is probably because of the POSIX standard:


TLDR: Although POSIX asks all implementers to support all characters in env names except `=`, POSIX standard says standard programs should only use `[a-zA-Z_][\w-]*\w$`.

> Environment variable names used by the utilities in the Shell and Utilities volume of
POSIX.1-2024 consist solely of uppercase letters, digits, and the <underscore> ('_') from the
characters defined in Table 6-1 (on page 117) and do not begin with a digit. Other characters, and
byte sequences that do not form valid characters, may be permitted by an implementation;
applications shall tolerate the presence of such names. Uppercase and lowercase letters shall
retain their unique identities and shall not be folded together. The name space of environment
variable names containing lowercase letters is reserved for applications. Applications can define
any environment variables with names from this name space without modifying the behavior of
the standard utilities.
>
> --<cite>1003.1-2024 - IEEE/Open Group Standard for Information Technology--Portable Operating System Interface (POSIX™) Base Specifications, Issue 8</cite>

</details>

Secondly I think your syntax has some limitations and I think there are arguably better alternatives:

- If one rule name shares the same prefix as another rule, there is ambiguity on which rule the config is actually referring to.
- If the backing program actually needs args in a "key=value" format, there is no unambiguous way to do that with this syntax.
- If the backing program just need a flag with no parameters, like "--verbose", there is no way to pass it.

There is already pretty accepted ways to do solve this problem in the programming community: Compiler drivers have need to pass on arguments to build stages while accepting arguments of their own all the time.

GCC passes options with `-Wl,` and then all options separated by comma, so in your case would be `--config set_prokka=--kindom,archaea,--locustag,XXX` and so on. There is no ambiguity here. There is currently no way to pass a comma verbatim in GCC AFAIK but it would not be too hard to implement an escaping mechanism if you have that need.

Rustc does very similar things except use spaces to separate.

**Comment by cmkobel** [comment]
I'm not asking to add the passthrough argument feature - Just to allow a hyphen in the key. The posix standard definition is exactly what I'm asking for. This is another argument why this should be implemented.

You bring up several points which I have already taken into consideration:

> If one rule name shares the same prefix as another rule, there is ambiguity on which rule the config is actually referring to.

In my [implementation](https://github.com/cmkobel/CompareM2/blob/9245cd6ae78fbbb58da6b87e69522ded24d9c982/workflow/rules/sample_annotation.smk#L71), the parameters are unpacked with unpack_parameters("<rule>"). Since you can manually specify the rule name, there is no conflict when several rules have the same prefix. 

> If the backing program actually needs args in a "key=value" format, there is no unambiguous way to do that with this syntax.

The example that I bring in my post is just a practical use case, to help explain why I think hyphens in config keys are empowering. One can easily [implement](https://github.com/cmkobel/CompareM2/blob/9245cd6ae78fbbb58da6b87e69522ded24d9c982/workflow/Snakefile#L72-L97) a way to define a specific delimiter for each rule:
```
    params:
        unpack_parameters("<rule>", "<delimiter>")
```
So if you wanted to go for equal-sign or even comma-sign (your GCC example), you could write that in the delimiter, and the snakemake config or command line config would look just the same. E.g. `unpack_parameters("GCC", ",")

> If the backing program just need a flag with no parameters, like "--verbose", there is no way to pass it.

This is done trivially by passing an empty string ([as I already explain to my users](https://comparem2.readthedocs.io/en/latest/20%20usage/#passthrough-arguments)):
_" [...] In some cases, command line options are flags, meaning that they need no parameter value. In this case, an empty string can be given as parameter value: `comparem2 --until set_prokka--rfam=""` "_

> I think there are arguably better alternatives:

Can you please elaborate? The only solution I've seen is that Snakemake pipeline developers write [complete argparse based command line parsers](https://github.com/rhysnewell/aviary/blob/387ac6db734d20265086fb307aa5bcbcfe087592/aviary/aviary.py#L125). This is error prone, time consuming and unnecessary when considering that Snakemake already has a great config system.

When deciding whether to allow hyphens in config keys given on the command line (which is the only thing I'm asking for), please take into account the fact that it will not have any negative or adverse effects on Snakemake, but will only make better able to solve challenges in the field.

Best, Carl

**Comment by cmkobel** [comment]
Fixed with https://github.com/snakemake/snakemake/pull/2998
